### PR TITLE
refactor: POS workflow

### PR DESCRIPTION
**Pull request by nextchamp-saqib**
## Goal
Point of Sale (online) will now create POS Invoices. These invoices are submitted faster than sales invoice as they don't hit Stock Ledger or General Ledger. These invoices will get merged into a consolidated Sales Invoices on POS Closing (or manually) based on customer.  Available Stock will be validated by recent POS Invoices and latest stock ledger entry.

## Roadmap
- [x] POS Invoices
- [x] POS Invoices can be merged into Sales Invoice
- [x] Auto merge on POS Closing
  - [x] Auto fetch transaction's details in POS Closing Voucher
- [x] Consolidated Credit Note for POS Returns
- [x] New UI
  - [x] Item Selector
     - [x] Fix item search and category selection
  - [x] Item Details
     - [x] Fix auto serial selection button for non serialised item
  - [x] Item Cart
  - [x] Payments
     - [x] Loyalty Points as a payment method
     - [x] Email Receipt / Print Receipt
  - [x] Past Orders Section
     - [x] Load past pos invoice in the UI. Can be submitted or draft
     - [x] Process Returns
- [x] POS Print Format
- [ ] Tests
- [ ] Docs

#### Item Selector

Screenshot 2020-04-12 at 3 20 08 PM

#### Item Details

Screenshot 2020-04-12 at 3 11 43 PM

#### Payments

Screenshot 2020-04-12 at 3 12 20 PM

#### Past Order List

Screenshot 2020-04-19 at 12 55 06 PM

### CODE REVIEW

**Review by nabinhait — CHANGES_REQUESTED** [review]
Fix conflict

File: erpnext/public/js/utils/serial_no_batch_selector.js [hunk 1]
```diff
@@ -74,31 +73,39 @@ erpnext.SerialNoBatchSelector = Class.extend({
...
 						let auto_fetched_serial_numbers = data.message;
 						let records_length = auto_fetched_serial_numbers.length;
+						if (!records_length) {
+							const warehouse = me.dialog.fields_dict.warehouse.get_value().bold();
+							frappe.msgprint(__(`Serial numbers unavailable for Item ${me.item.item_code.bold()} 
+								under warehouse ${warehouse}. Please try changing warehouse.`));
+						}
 						if (records_length < qty) {
-							frappe.msgprint(`Fetched only ${records_length} serial numbers.`);
+							frappe.msgprint(`Fetched only ${records_length} avialable serial numbers.`);
```
Comment (mhbu50) [inline erpnext/public/js/utils/serial_no_batch_selector.js]: Make the text translatable

File: erpnext/selling/page/point_of_sale_beta/pos_payment.js [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+		)
+	}
+
+	show_totals(doc) {
+		if (!doc) doc = this.events.get_doc();
+		const paid_amount = doc.paid_amount;
+		const remaining = doc.grand_total - doc.paid_amount;
+		const change = doc.change_amount;
+		const currency = doc.currency
+		const label = change ? 'Change' : 'To Be Paid';
```
Comment (mhbu50) [inline erpnext/selling/page/point_of_sale_beta/pos_payment.js]: Make label translateable

File: erpnext/accounts/page/pos/pos.js [hunk 3]
```diff
@@ -1766,7 +1766,7 @@ erpnext.pos.PointOfSale = erpnext.taxes_and_totals.extend({
 
 		if (this.si_docs.length || this.email_queue_list || this.customers_list) {
 			frappe.call({
-				method: "erpnext.accounts.doctype.sales_invoice.pos.make_invoice",
+				method: "erpnext.accounts.page.pos.pos.make_invoice",
```
Comment (rohitwaghchaure) [inline erpnext/accounts/page/pos/pos.js]: remove offline POS

File: erpnext/patches/v12_0/rename_pos_closing_doctype.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+		if not frappe.db.exists("DocType", "POS Closing Entry"):
+			frappe.rename_doc('DocType', 'POS Closing Voucher', 'POS Closing Entry', force=True)
+		
+		if not frappe.db.exists('DocType', 'POS Closing Entry Taxes'):
+			frappe.rename_doc('DocType', 'POS Closing Voucher Taxes', 'POS Closing Entry Taxes', force=True)
+		
+		if not frappe.db.exists('DocType', 'POS Closing Voucher Details'):
+			frappe.rename_doc('DocType', 'POS Closing Voucher Details', 'POS Closing Entry Details', force=True)
+
+		frappe.reload_doc('Selling', 'doctype', 'POS Closing Entry')
```
Comment (nabinhait) [inline erpnext/patches/v12_0/rename_pos_closing_doctype.py]: Move these documents to accounts module

File: erpnext/patches/v13_0/replace_pos_payment_mode_table.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+		if payments:
+			for payment_mode in payments:
+				pos_payment_method = frappe.new_doc("POS Payment Method")
+				pos_payment_method.idx = payment_mode.idx
+				pos_payment_method.default = payment_mode.default
+				pos_payment_method.mode_of_payment = payment_mode.mode_of_payment
+				pos_payment_method.parent = payment_mode.parent
+				pos_payment_method.parentfield = payment_mode.parentfield
+				pos_payment_method.parenttype = payment_mode.parenttype
+				pos_payment_method.save()
```
Comment (nabinhait) [inline erpnext/patches/v13_0/replace_pos_payment_mode_table.py]: db_insert()

File: erpnext/selling/doctype/pos_closing_entry/pos_closing_entry.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+from erpnext.controllers.taxes_and_totals import get_itemised_tax_breakup_data
+from erpnext.selling.doctype.pos_invoice_merge_log.pos_invoice_merge_log import merge_pos_invoices
+
+class POSClosingEntry(Document):
+	def validate(self):
+		user = frappe.get_all('POS Closing Entry',
+			filters = { 'user': self.user, 'docstatus': 1 },
+			or_filters = {
+					'period_start_date': ('between', [self.period_start_date, self.period_end_date]),
+					'period_end_date': ('between', [self.period_start_date, self.period_end_date])
```
Comment (nabinhait) [inline erpnext/selling/doctype/pos_closing_entry/pos_closing_entry.py]: or self.period_start_date between period_start_date and period_end_date

File: erpnext/selling/doctype/pos_closing_entry_details/pos_closing_entry_details.json [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+   "options": "company:company_currency",
+   "reqd": 1
+  }
+ ],
+ "istable": 1,
+ "links": [],
+ "modified": "2020-04-28 16:43:11.448881",
+ "modified_by": "Administrator",
+ "module": "Selling",
+ "name": "POS Closing Entry Details",
```
Comment (nabinhait) [inline erpnext/selling/doctype/pos_closing_entry_details/pos_closing_entry_details.json]: Name should be singular "POS Closing Entry Detail"

File: erpnext/selling/doctype/pos_invoice/pos_invoice.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+		self.validate_debit_to_acc()
+		self.validate_write_off_account()
+		self.validate_change_amount()
+		self.validate_item_cost_centers()
+		self.validate_serialised_or_batched_item()
+		self.validate_stock_availablility()
+		self.validate_return_items()
+		self.set_status()
+		self.set_account_for_mode_of_payment()
+		if cint(self.is_pos):
```
Comment (nabinhait) [inline erpnext/selling/doctype/pos_invoice/pos_invoice.py]: Put the below code inside a function

File: erpnext/selling/doctype/pos_opening_entry_details/pos_opening_entry_details.json [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+   "options": "company:company_currency",
+   "reqd": 1
+  }
+ ],
+ "istable": 1,
+ "links": [],
+ "modified": "2020-04-28 16:58:02.125902",
+ "modified_by": "Administrator",
+ "module": "Selling",
+ "name": "POS Opening Entry Details",
```
Comment (nabinhait) [inline erpnext/selling/doctype/pos_opening_entry_details/pos_opening_entry_details.json]: POS Opening Entry Detail

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Review by nabinhait — CHANGES_REQUESTED** [review]
- Error during migrate:
1.
```
Executing execute:frappe.reload_doc("selling", "doctype", "POS Payment Method") #2020-05-28 in develop (_01c2b3de3a80a4dd)
/Users/nabinhait/projects/develop-bench/apps/erpnext/erpnext/selling/doctype/pos_payment_method/pos_payment_method.json missing
```
2. 
```
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/utils/bench_helper.py", line 97, in <module>
    main()
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/utils/bench_helper.py", line 18, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/Users/nabinhait/projects/develop-bench/env/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/nabinhait/projects/develop-bench/env/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/nabinhait/projects/develop-bench/env/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/nabinhait/projects/develop-bench/env/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/nabinhait/projects/develop-bench/env/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/nabinhait/projects/develop-bench/env/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/nabinhait/projects/develop-bench/env/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/commands/__init__.py", line 25, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/commands/site.py", line 260, in migrate
    migrate(context.verbose, rebuild_website=rebuild_website, skip_failing=skip_failing)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/migrate.py", line 70, in migrate
    frappe.model.sync.sync_all(verbose=verbose)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/sync.py", line 19, in sync_all
    sync_for(app, force, verbose=verbose, reset_permissions=reset_permissions)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/sync.py", line 70, in sync_for
    reset_permissions=reset_permissions, for_sync=True)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/modules/import_file.py", line 70, in import_file_by_path
    ignore_version=ignore_version, reset_permissions=reset_permissions)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/modules/import_file.py", line 143, in import_doc
    doc.insert()
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/document.py", line 264, in insert
    self.run_post_save_methods()
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/document.py", line 948, in run_post_save_methods
    self.run_method("on_update")
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/document.py", line 815, in run_method
    out = Document.hook(fn)(self, *args, **kwargs)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/document.py", line 1101, in composer
    return composed(self, method, *args, **kwargs)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/document.py", line 1084, in runner
    add_to_return_value(self, fn(self, *args, **kwargs))
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/model/document.py", line 809, in <lambda>
    fn = lambda self, *args, **kwargs: getattr(self, method)(*args, **kwargs)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/core/doctype/doctype/doctype.py", line 291, in on_update
    self.run_module_method("on_doctype_update")
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/core/doctype/doctype/doctype.py", line 357, in run_module_method
    module = load_doctype_module(self.name, self.module)
  File "/Users/nabinhait/projects/develop-bench/apps/frappe/frappe/modules/utils.py", line 206, in load_doctype_module
    raise ImportError('Module import failed for {0} ({1})'.format(doctype, module_name + ' Error: ' + str(e)))
ImportError: Module import failed for POS Closing Entry (erpnext.accounts.doctype.pos_closing_entry.pos_closing_entry Error: No module named 'erpnext.selling.doctype.pos_invoice_merge_log.pos_invoice_merge_log')
```

**Comment by olamide2** [comment]
Hi @nextchamp-saqib 

While we are doing this can we work on item variants? 

1. I noticed that their is a problem selecting items with variants into the shopping cart. Error log is as follows 


erroimage

Also can we change the workflow for selecting items with variants to follow the sequence in the gif image below ? This is assuming it is not been scanned in with barcode. 

pos2

**Comment by nextchamp-saqib** [comment]
@olamide2 Send me the item form view for the item with variants.

**Comment by olamide2** [comment]
> @olamide2 Send me the item form view for the item with variants.

The item view in ERPNext or in the app I used as an example above?

**Comment by nextchamp-saqib** [comment]
> > @olamide2 Send me the item form view for the item with variants.
> 
> The item view in ERPNext or in the app I used as an example above?

In ERPNext

**Comment by nextchamp-saqib** [comment]
@olamide2  The item description is faulty. Please send only the item description screenshot.

**Comment by olamide2** [comment]
image
image
image

**Comment by olamide2** [comment]
I do not think it I'd faulty,

It is auti generated by the system

**Comment by nextchamp-saqib** [comment]
@olamide2 I cannot replicate with the same configuration as yours. However I will make changes to account for such descriptions.

**Comment by olamide2** [comment]
I will test this.

Thank you

**Review by nabinhait — CHANGES_REQUESTED** [review]
- [x] POS link in desk page
- [x] Delete old pos and point-of-sale page
- [x] Me is not defined (onload of pos beta) - pos_controller.js - 517
- [x] /Users/nabinhait/projects/develop-bench/apps/erpnext/erpnext/selling/doctype/pos_payment_method/pos_payment_method.json missing in migrate
- [x] In recent orders page, on change of order type, clear selected form view
- [x] “Error: Data missing in table: Items” coming from recent orders page, on click of “save as draft”
- [x] “jinja2.exceptions.TemplateNotFound: erpnext/selling/doctype/pos_closing_entry/closing_voucher_details.html” on click of “Close the POS”
- [ ] Global search using "pos" term --- No provision in framework for this
- [ ] “Tax template” naming and style -- Fixed naming
- [ ] POS Opening Entry - “01-06-2020 un11fin11” in period start date field -- Unable to replicate
- [x] On saving POS Closing Entry - “Payment Reconciliation” table is getting blank

**Comment by olamide2** [comment]
Hello Guys

Any update on this? We are all waiting for it to be merged so we can start adding our own pull requests to this 

We fully intend to make this the best PoS in the market. No joke.

**Comment by olamide2** [comment]
Opening and Closing Entry

When carrying out the "Close the PoS" function I noticed that the system does not show the difference amount when the closing amount is keyed in

image

It also does not show the opening amount as keyed in when the cashier started the day.

Kindly look into this 

Regards

**Comment by nextchamp-saqib** [comment]
Opening amount should be set while creating an opening entry,  which will be auto fetched while closing.

Difference amount will be calculated automatically based on keyed Closing amount.

**Comment by olamide2** [comment]
> Opening amount should be set while creating an opening entry, which will be auto fetched while closing.
> 
> Difference amount will be calculated automatically based on keyed Closing amount.

I know all this already, I am telling you that the system is not behaving that way.

I also pasted an image where the difference amount is not calculating see the image above, the closing amount and the Expected Amount are not the same yet the difference amount is zero.

Also in the case above I keyed in an opening amount and yet it shows zero above.

Please check

**Comment by olamide2** [comment]
Another issues you might want to look at. This is a process issue.

Consider a case where the company does not allow negative stock.

I see that at Point of Sale the cashier gets a warning when an item that will go into negative is in the shopping cart.

However the sales transaction is still allowed to be completed.

However this negative stock situation will not allow the PoS closing entry to be concluded at end of day. It will ask that this negative situation be corrected first.

In my opinion if the organisation chooses not to allow negative stock then the transaction should not go through at PoS. It will be troublesome for the end of day team to be the one to worry about negative stock.

What do you think?

**Comment by olamide2** [comment]
By the way please do not remove offline PoS

The two must still  be able to live side by side

**Review by nabinhait — CHANGES_REQUESTED** [review]
Travis is failing

**Comment by Olawale1** [comment]
> In my opinion if the organisation chooses not to allow negative stock then the transaction should not go through at PoS. It will be troublesome for the end of day team to be the one to worry about negative stock.
> 
> What do you think?

Hi @olamide2 

I think it might be better for the end of day team to deal with it rather than the cashier in front of queueing customers

Cheers

**Review by nabinhait — APPROVED** [review]

**Comment by muchai** [comment]
Hi, hope this addresses this bug: https://github.com/frappe/erpnext/issues/20889

**Comment by ptdas** [comment]
Hi,

please add function to allow print before submit just like the older version... 

that will help customer that need a pre print receipt for task checking (for F&B)

Thanks

**Comment by dexterhussain** [comment]
anyway to add sale person in this ?

**Comment by Botanium** [comment]
Do we have a weight scale feature in the POS?

**Comment by Labyrinth-00** [comment]
the item selector is not showing the full price.  Just the integer part of the decimal price.

**Comment by RiyadhAltayib** [comment]
Hello Frappe team

I'm so excited for the release of the new UI and After testing the POS; I've got some points I wish to raise to your kind consideration. I hope this would be the right place to write my request.

1- **Partial payment through POS view:**
[Case#1] combined payment that is paid partially by cash and reminder paid by other mode of payment.
[Case#2] partial payment, and upon confirming the invoice (the user can re-open same invoice from POS view) to be able to update collection of remaining payment.

2- **Payment on Credit through POS view:**
[Case#1] customer has loyalty points or a credit which will be deducted fully from upon invoice confirmation.
[Case#2] customer has loyalty points or a credit which will be deducted partially from and reminder paid by other mode of payment.
[Case#3] customer is provided with the completed invoice with zero payment and is allowed to pay later.

3- **Print Invoice (before submit - with note of Unpaid) through POS view**

4- **write-off invoice (completely or partially) from POS view**

5- **Ability to edit and customize the POS view** 
To add additional fields such as (barcode reader for quick item select).

**Pull request merged**